From be193f3a97dce50bad1cf6100aa662c2f6ba57f7 Mon Sep 17 00:00:00 2001
From: Michael Matloob <matloob@golang.org>
Date: Fri, 3 Apr 2026 15:03:36 -0400
Subject: [PATCH] cmd/go: use fsys.ReadDir for IsStandardPackage

Backport of https://github.com/golang/go/commit/be193f3a97dce50bad1cf6100aa662c2f6ba57f7

FIPS140 crypto files will be bound into the virtual filesystem using the
fsys package. So IsStandardPackage needs to use fsys.ReadDir to check
that the fips140 packages are standard packages rather than os.ReadDir
because os.ReadDir doesn't know about the overlay.

Fixes https://github.com/golang/go/issues/73649
---
 src/cmd/go/internal/modindex/read.go | 4 ++--
 src/go/build/build.go               | 2 +-
 src/internal/goroot/gc.go            | 6 +++---
 3 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/src/cmd/go/internal/modindex/read.go b/src/cmd/go/internal/modindex/read.go
index 399e89eca3..a31e09d233 100644
--- a/src/cmd/go/internal/modindex/read.go
+++ b/src/cmd/go/internal/modindex/read.go
@@ -680,7 +680,7 @@ func (rp *IndexPackage) Import(bctxt build.Context, mode build.ImportMode) (p *b
 // and otherwise falling back to internal/goroot.IsStandardPackage
 func IsStandardPackage(goroot_, compiler, path string) bool {
 	if !enabled || compiler != "gc" {
-		return goroot.IsStandardPackage(goroot_, compiler, path)
+		return goroot.IsStandardPackage(fsys.ReadDir, goroot_, compiler, path)
 	}
 
 	reldir := filepath.FromSlash(path) // relative dir path in module index for package
@@ -695,7 +695,7 @@ func IsStandardPackage(goroot_, compiler, path string) bool {
 	} else if errors.Is(err, ErrNotIndexed) {
 		// Fall back because package isn't indexable. (Probably because
 		// a file was modified recently)
-		return goroot.IsStandardPackage(goroot_, compiler, path)
+		return goroot.IsStandardPackage(fsys.ReadDir, goroot_, compiler, path)
 	}
 	return false
 }
diff --git a/src/go/build/build.go b/src/go/build/build.go
index 157461d1af..4c6be413ab 100644
--- a/src/go/build/build.go
+++ b/src/go/build/build.go
@@ -746,7 +746,7 @@ func (ctxt *Context) Import(path string, srcDir string, mode ImportMode) (*Packa
 				}
 				tried.goroot = dir
 			}
-			if ctxt.Compiler == "gccgo" && goroot.IsStandardPackage(ctxt.GOROOT, ctxt.Compiler, path) {
+			if ctxt.Compiler == "gccgo" && goroot.IsStandardPackage(os.ReadDir, ctxt.GOROOT, ctxt.Compiler, path) {
 				// TODO(bcmills): Setting p.Dir here is misleading, because gccgo
 				// doesn't actually load its standard-library packages from this
 				// directory. See if we can leave it unset.
diff --git a/src/internal/goroot/gc.go b/src/internal/goroot/gc.go
index 534ad57e70..d2581b0424 100644
--- a/src/internal/goroot/gc.go
+++ b/src/internal/goroot/gc.go
@@ -15,12 +15,12 @@ import (
 )
 
 // IsStandardPackage reports whether path is a standard package,
-// given goroot and compiler.
-func IsStandardPackage(goroot, compiler, path string) bool {
+// given goroot and compiler. readDir accepts OS filesystem paths.
+func IsStandardPackage(readDir func(string) ([]os.DirEntry, error), goroot, compiler, path string) bool {
 	switch compiler {
 	case "gc":
 		dir := filepath.Join(goroot, "src", path)
-		dirents, err := os.ReadDir(dir)
+		dirents, err := readDir(dir)
 		if err != nil {
 			return false
 		}
